File: resource_server/src/main/java/com/commerce/backend/service/CartService.java
package com.commerce.backend.service;

import com.commerce.backend.model.entity.Cart;
import com.commerce.backend.model.request.cart.ConfirmCartRequest;
import com.commerce.backend.model.response.cart.CartResponse;


public interface CartService {
    CartResponse addToCart(Long id, Integer amount);

    CartResponse incrementCartItem(Long cartItemId, Integer amount);

    CartResponse decrementCartItem(Long cartItemId, Integer amount);

    CartResponse fetchCart();

    CartResponse removeFromCart(Long id);

    boolean confirmCart(ConfirmCartRequest confirmCartRequest);

    Cart getCart();

    void saveCart(Cart cart);

    void emptyCart();

    Cart calculatePrice(Cart cart);

}

File: client/src/app/services/cart.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { config } from '../../config/local';
import { Cart } from '../store/model';


@Injectable()
export class CartService {

  url = `${config.apiUrl}/api/cart`;

  constructor(private httpClient: HttpClient) {
  }

  getCart() {
    return this.httpClient.get<Cart>(this.url);
  }

  postCart(productVariantId: number, amount: string) {
    return this.httpClient.post<Cart>(this.url, {
      productVariantId,
      amount
    });
  }

  incrementCartItem(cartItemId: number, amount: string) {
    return this.httpClient.post<Cart>(`${this.url}/increment`, {
      cartItemId,
      amount: Number(amount)
    });
  }

  decrementCartItem(cartItemId: number, amount: string) {
    return this.httpClient.post<Cart>(`${this.url}/decrement`, {
      cartItemId,
      amount: Number(amount)
    });
  }

  removeFromCart(id: number) {
    return this.httpClient.post<Cart>(this.url + '/remove', {
      cartItemId: id
    });
  }

  confirmCart(cart: Cart) {
    return this.httpClient.post(this.url + '/confirm', cart);
  }

  applyDiscount(code: string) {
    return this.httpClient.post<Cart>(`${this.url}/discount`, {
      code
    });
  }

  emptyCart() {
    return this.httpClient.delete(this.url);
  }


}
